File: isValidSudoku.py
#Hashset strategy
#Time complexity -> O(n^2) nested for loops
#Space complexity -> O(n^2) in memory bc it store sets for each row, column, and square. Total of 27 sets (9 column sets, 9 row sets, and 9 square sets) n=9, total of 81 tiles
from typing import List
from collections import defaultdict #provides default value for nonexistent key, avoids KeyError
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        #sets because we only need to figure out if there are any duplicates
        cols = defaultdict(set)
        rows = defaultdict(set)
        squares = defaultdict(set) #key = (row/3,column/3)

        for r in range(9):
            for c in range(9):
                if board[r][c] == ".":
                    continue
                if(board[r][c] in rows[r]
                or board[r][c] in cols[c]
                or board[r][c] in squares[r//3, column//3]):
                    return False

                cols[c].add(board[r][c])
                rows[r].add(board[r][c])
                squares[r//3,c//3].add(board[r][c])
        
        return True

File: isValidSudoku.cpp
//HashSet
//Time complexity -> O(n^2)
//Space Complexity -> O(n^2)
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <map> //unordered_map or map can be used for storing sets of values for the squares (map = sorted, unordered_map = not-sorted)
using namespace std;

class Solution{
    bool isValidSudoku(vector<vector<char>>& board){
        unordered_map<int, unordered_set<char>> rows, cols;
        map<pair<int,int>, unordered_set<char>> squares;

        for(int r = 0; r < 9; r++){
            for(int c = 0; c < 9; c++){
                if (board[r][c] == '.'){
                    continue;
                }

                pair<int,int> squareKey = {r/3, c/3};

                if(rows[r].count(board[r][c]) ||
                cols[c].count(board[r][c]) ||
                squares[squareKey].count(board[r][c])){
                    return false;
                }

                rows[r].insert(board[r][c]);
                cols[c].insert(board[r][c]);
                squares[squareKey].insert(board[r][c]);
            }
        }
        return true;
    }
};